[Utils] add update-verify-tests.py

hnrklssn opened the pull request:
Adds a python script to automatically take output from a failed clang -verify test and update the test case(s) to expect the new behaviour.

hnrklssn committed:
[Utils] add update-verify-tests.py

Adds a python script to automatically take output from a failed clang
-verify test and update the test case(s) to expect the new behaviour.

nhaehnle reviewed (commented):
This is a nice addition, but I think it should follow the conventions established by the existing update_*_test_checks.py scripts as much as possible, at least:

* Ability to parse RUN lines, re-execute them autonomously, and modify test files in place based on their output
* Proper argument parsing via the argparse library
* Name of the script itself

You should integrate with the UpdateTestChecks/common.py infrastructure as much as possible. At a minimum, the collection of test cases to update, the RUN line parsing, and stuff like `common.invoke_tool` should almost certainly be shared. This may require some minor updates to that common infrastructure, but that's a good thing.

The actual generating and updating of CHECK lines is probably sufficiently different that sharing might not make sense. That said, I wonder if this script can be generalized to automatically generate CHECK lines for other tools that produce diagnostic output.

hnrklssn commented:
> This is a nice addition, but I think it should follow the conventions established by the existing update_*_test_checks.py scripts as much as possible, at least:
> 
>     * Ability to parse RUN lines, re-execute them autonomously, and modify test files in place based on their output
> 
>     * Proper argument parsing via the argparse library
> 
>     * Name of the script itself
> 
> 
> You should integrate with the UpdateTestChecks/common.py infrastructure as much as possible. At a minimum, the collection of test cases to update, the RUN line parsing, and stuff like `common.invoke_tool` should almost certainly be shared. This may require some minor updates to that common infrastructure, but that's a good thing.

I did consider this at first - here's my reasoning why I didn't, please let me know what you think:

I think that the added complexity of RUN line parsing + `common.invoke_tool` would cost more than it's worth for this tool. It's already a reimplementation of something that `lit` does excellently. `lit` also has automatic test discovery, instead of having to manually provide each test case to the script. Any test case relying on RUN lines is going to be tested with `lit` anyways, so it's not like we're adding any extra coupling.

The `update_*_test_checks.py` scripts use `common.invoke_tool` to prevent the output from being piped into `FileCheck`, but this script uses the raw output of the `-verify` flag (which is rarely if ever piped into anything), so invoking the lines manually doesn't provide the same extra value.

If anything I think `lit` should have a plugin system that allows a plugin to intercept the test invocations and control how RUN lines are executed. I didn't implement that now though, because I didn't want to expand the scope of this patch too much.

> The actual generating and updating of CHECK lines is probably sufficiently different that sharing might not make sense. That said, I wonder if this script can be generalized to automatically generate CHECK lines for other tools that produce diagnostic output.

This script relies on the format of the output of the `-verify` flag to get the ground truth of which diagnostics are mismatching, and how. FileCheck doesn't provide as comprehensive information because it's much more flexible, so it'd be harder for it to confidently say "this check doesn't match anything and this check should be inserted on this line". E.g. this script doesn't even try to handle checks with regexes, while any FileCheck check can contain a regex unless they specify `LITERAL`.

For the checks themselves I think they are sufficiently different from FileCheck checks that it'd be hard to generalise the output: multiple `expected-*` checks can point to the same line, and at the same time we have no information about the order the diagnostics are actually emitted to the use, e.g. a later diagnostic can point to an earlier line.

hnrklssn committed:
[Utils] Add testing for update-verify-tests, fix bugs

Fixes various line offset bugs. Fixes a bug where whitespace could be
emitted between 'expected-[category]' and the line location specifier,
which clang does not parse.

Now handles the cases where the file does not emit any diagnostics, or
previously contained 'expected-no-diagnostics'.
Also does further work to minimise diffs by replacing diagnostics checks
with a new one, even if the previous check was in a location that the
script would not normally emit. This is useful for the case where a
check is on the same line as the emitted diagnostic, and the diagnostic
message changes. Instead of removing the previous diagnostic and
inserting a new one with @+1, the old diagnostic text is replaced with
the new one.

hnrklssn committed:
[Utils] Add custom prefix support to update-verify-tests

Custom prefixes can now be provided with --prefix. The default remains
'expected'.

hnrklssn commented:
Added tests and updated the script. Instead of only being a free-standing script it's now (along with the UTC scripts) integrated into llvm-lit to automatically detect which script can update a failing test, using the new flag `--update-tests`.

nhaehnle commented on clang/test/utils/update-verify-tests/Inputs/duplicate-diag.c:
```diff
@@ -0,0 +1,10 @@
+// RUN: %clang_cc1 -verify %s
+// RUN: diff %s %s.expected
```
The way I understand it, the update-verify-tests plugin in lit will overwrite this source file, which may end up overwriting this RUN line if there is a bug. Basically, there is a weird circularity here.

Therefore, it seems to me that the RUN lines for the `diff` against expected ought to be in the `lit-plugin.test` file instead.

hnrklssn replied:
The files in the `Inputs` directory are not discovered by lit, so this doesn't actually run the RUN lines. Instead a copy is made, and a new lit instance is launched to test the run lines.

This is in `duplicate-diag.test`, the RUN lines of `duplicate-diag.c` are not executed at all for this case because it tests the `update-verify-tests.py` script:
```
# RUN: cp %S/Inputs/duplicate-diag.c %t.c && not %clang_cc1 -verify %t.c 2>&1 | %update-verify-tests
# RUN: diff -u %S/Inputs/duplicate-diag.c.expected %t.c
# RUN: %clang_cc1 -verify %t.c
```

This is in `lit-plugin.test`, which copies the files to a separate directory and launches llvm-lit there to test the lit integration:
```
# RUN: cp %S/Inputs/*.c %S/LitTests
# RUN: cp %S/Inputs/*.c.expected %S/LitTests
# RUN: not %{lit} %S/LitTests
# RUN: not %{lit} %S/LitTests --update-tests
# RUN: %{lit} %S/LitTests
```

nhaehnle replied:
Right, it's all a little mind-bending, but my point is that since you're trying to test whether the invocation of `not %{lit} %S/LitTests --update-tests` in `lit-plugin.test` updated the tests to look like the `.expected` file, that invocation of `diff` should *also* be in `lit-plugin.test`.

You can't have the correctness check for a test that modifies a file be in the same file that is being modified. It has to live outside.

nhaehnle reviewed (commented):
I think the integration into lit is okay. The diff updater seems neat, but I think it could be taken into a separate change, see also the inline comment. In general, I think this PR really does three separate things, that ought to be in separate changes:

* Add update-verify-tests.py
* Add a way to update tests from within llvm-lit
* Add the diff-test-updater to llvm-lit

hnrklssn commented:
> I think the integration into lit is okay. The diff updater seems neat, but I think it could be taken into a separate change, see also the inline comment. In general, I think this PR really does three separate things, that ought to be in separate changes:
> 
>     * Add update-verify-tests.py
> 
>     * Add a way to update tests from within llvm-lit
> 
>     * Add the diff-test-updater to llvm-lit

Makes sense, I'll separate the changes into 3 PRs.

hnrklssn committed:
[Utils] Separate update-verify-tests script from core

The script no longer attempts to be RUN line aware. Feeding raw llvm-lit
output into it _may_ still work, but given a large enough number of
test cases to update, the risk that some test case uses unsupported
features grows pretty large. This will instead be handled by adding a
lit integration to automatically invoke the core for each failing test
output. The lit integration will be landed as a separate change.

nhaehnle reviewed (approved):
I didn't go over the core.py implementation with a comb, but this looks reasonable to me. Thanks!

jurahul commented:
Looks like these tests are failing in CI: see https://buildkite.com/llvm-project/github-pull-requests/builds/100644#0191ec9f-441f-4299-ad47-a9318fbae7a8

nico commented:
This breaks tests on windows: http://45.33.8.238/win/94083/step_6.txt

Please take a look and revert for now if it takes a while to fix.

jurahul commented:
Also note that this was also detected by CI prior to merging the PR.

jurahul commented:
Reverted in https://github.com/llvm/llvm-project/pull/108630

hnrklssn commented:
Very sorry about the CI oversight on my part! Thanks for reverting, I was in a meeting.
Looks like I was bitten by Windows line endings again...
